compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+
+  gen += "void chpl_mli_server_set_conf(void)";
+  gen += scope_begin;
+  gen += "chpl_server_conf.debug=";
+  gen += this->injectDebugPrintlines ? "1" : "0";
+  gen += ";\n";
+  gen += scope_end;
+  
+  return gen;
+}
+
+void MLIContext::emitServerPrelude(void) {
+  std::string gen;
+
+  // TODO: Trying to work around the Travis malloc complaint.
+  gen += "#define CHPL_MLI_IS_SERVER_\n";
+  gen += this->genHeaderInc("chpl_mli_marshalling.c");
+  gen += this->genNote("We use Makefile magic to make this visible!");
+  gen += this->genHeaderInc("mli_server_runtime.c");
+  gen += this->genHeaderInc("_main.c");
+  gen += "\n";
+  
+  // The server will call this function to set globals appropriately.
+  gen += this->genFuncToSetServerGlobals();
+  gen += "\n";
+
+  this->setOutputAndWrite(&this->fiServerBundle, gen);
+
+  return;
+}
+
+void MLIContext::emitMarshalRoutines(void) {
+  std::map<Type*, int64_t>::iterator i;
+  std::string gen;
+
+  // Bunch of stuff included in generated header as bridge code.
+  gen += this->genHeaderInc("stdlib.h", true);
+  gen += this->genHeaderInc("stdio.h", true);
+  gen += this->genHeaderInc("zmq.h", true);
+  gen += this->genHeaderInc("chpl__header.h");
+  gen += this->genNote("We use Makefile magic to make this visible!");
+  gen += this->genHeaderInc("mli_common_code.c");
+  gen += "\n";
+
+  for (i = this->typeMap.begin(); i != this->typeMap.end(); ++i) {
+    if (this->injectDebugPrintlines) {
+      std::string tpn = this->genTypeName(i->first);
+      gen += this->genComment(tpn.c_str());
+    }
+
+    gen += this->genMarshalPushRoutine(i->first);
+    gen += this->genMarshalPullRoutine(i->first);
+  }
+
+  this->setOutputAndWrite(&this->fiMarshalling, gen);
+
+  return;
+}
+
+//
+//
+//
+std::string MLIContext::genMarshalBodyPrimitiveScalar(Type* t, bool out) {
+  std::string gen;
+
+  // On pack, target buffer is input parameter. On unpack, the temporary.
+  const char* target = out ? "obj" : "result";
+
+  // Move the raw bytes of the type to/from the wire.
+  gen += "skt_err=";
+  gen += this->genSocketCall("skt", target, out);
+
+  // Generate a null frame in the opposite direction for the ACK.
+  gen += this->genSocketCall("skt", NULL, not out);
+
+  return gen;
+}
+
+//
+//
+//
+std::string MLIContext::genMarshalBodyStringC(Type* t, bool out) {
+  const char* target = out ? "obj" : "result";
+  std::string gen;
+
+  // Read the string length before pushing.
+  if (out) { gen += "bytes = strlen(obj);\n"; }
+
+  // Push/pull string length along with null frame for ACK.
+  gen += this->genSocketCall("skt", "bytes", out);
+  gen += this->genSocketCall("skt", NULL, not out);
+
+  if (out) {
+    // Next push this string across the wire.
+    gen += this->genSocketCall("skt", "obj", "bytes", out);
+  } else {
+    // Allocate buffer for string (include NULL terminator).
+    gen += "result = malloc(bytes + 1);\n";
+    gen += this->genTodo("Assert should shutdown server or return NULL.");
```

Comment:
As in the assert should be replaced by a shutdown call?  Or the assert function should shutdown the server?

Reply:
I just meant that if I'm going to use an assert, it should be an assert that also communicates to the server that it should shutdown, so that the server is always cleaned up.

I'm not sure how to handle internal malloc failure yet.

For strings specifically, the wise thing would probably be to return NULL to the user, but I haven't done that yet.

Reply:
It seems like you're trying to talk about two things in the same comment and that's confusing, at least to me.

Reply:
True. The real issue at hand here is how malloc failures should be handled.